docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+
+
+### The Type-Isolation barrier, and Client/Server send options
+One of the problems with getting proxies like this working is getting past the type-isolation barrier from ALCs. For example, if you want to use type `Foo`, loading `Foo` directly vs using an ALC generates what the runtime says are completely different types, even if the implementation is the same. 
+
+![](../img/ALCContextError.PNG)
+
+This happens for any non-simple types that attempt to cross the barrier, which is why API needs to use reflection to get the type of the ProxyObject instead of being able to grab the equivalent type in the user context.
+
+The type-isolation barrier becomes a larger issue for parameters and return types in particular, as when we pass a `Foo` type into the proxy, we need to transform it into the `ALC.Foo` type during the transition. There are a few ways we could do this:
+* Build proxies for each non-simple type that can communicate across the ALC boundaries
+* Make everything a shared type (This is not recommended)
+* Encode any passed objects and decode them on the other side, creating an effective "copy" of the type in the form `ALC.Foo`
+
+The first two have their own issues, as we would either need to figure out how to build those proxies on the fly from the IL-Generated `ProxyObject`/`ServerObject`, or lose out on the benefit of versioning compatibility. The third option is what `ServerObject` and `ClientObject` intend on setting a framework for; allowing for a (possibly interchangeable) connection framework to be used to send messages to and from the ALCs (or processes).
+
+There are a few ways we could implement the client/server encoding and messaging. The few that I've been looking at are:
+* Add a second `DispatchProxy` that can call across the barrier and keep our abstraction layer.
+* Using `BinaryFormatter` to serialize data across the boundary
+* Similar serialization with `DataContractSerializer`
+* Using gRPC to send objects across
+
+#### Second `DispatchProxy`
+Making the `ClientObject` a `DispatchProxy` object would work for making calls over ALCs, though not for out-of-process calls. It's not the perfect solution, as it would require two `Reflection.Invoke` calls (one from client to server, and one from server to target), but it would allow us to enforce our isolation policy from the custom `ServerObject` being able to mess with our return calls to make sure we were enforcing our constraints (see "ALC Unloading" below).
+
+#### Binary Serialization
+Using `BinaryFormatter`, types and their information can be serialized and sent across the ALC barrier, and deserialized using a class for decoding, which can then call the specific `TargetObject` methods.
+
+Benefits to using `BinaryFormatter`:
+* It is usable with any objects, as long as they are given the [Serializable] attribute.
+* There's no need to re-implement serialization, which would probably be out of scope for this project.
+
+Concerns with using `BinaryFormatter`:
```

Comment:
BinaryFormatter is deprecated technology that should not be used for any new code or projects.